pczt/src/orchard.rs

```diff
@@ -0,0 +1,524 @@
+    /// separate because they might need to be independently redacted. (For which role?)
+    pub(crate) rho: Option<[u8; 32]>,
+
+    /// The seed randomness for the note being spent.
+    ///
+    /// - This is set by the Constructor.
+    /// - This is required by the Prover.
+    pub(crate) rseed: Option<[u8; 32]>,
+
+    /// The full viewing key that received the note being spent.
+    ///
+    /// - This is set by the Updater.
+    /// - This is required by the Prover.
+    #[serde_as(as = "Option<[_; 96]>")]
+    pub(crate) fvk: Option<[u8; 96]>,
+
+    /// A witness from the note to the bundle's anchor.
+    ///
+    /// - This is set by the Updater.
+    /// - This is required by the Prover.
+    pub(crate) witness: Option<(u32, [[u8; 32]; 32])>,
+
+    /// The spend authorization randomizer.
+    ///
+    /// - This is chosen by the Constructor.
+    /// - This is required by the Signer for creating `spend_auth_sig`, and may be used to
+    ///   validate `rk`.
+    /// - After`zkproof` / `spend_auth_sig` has been set, this can be redacted.
+    pub(crate) alpha: Option<[u8; 32]>,
+
+    /// The ZIP 32 derivation path at which the spending key can be found for the note
+    /// being spent.
+    pub(crate) zip32_derivation: Option<Zip32Derivation>,
+
+    /// The spending key for this spent note, if it is a dummy note.
+    ///
+    /// - This is chosen by the Constructor.
+    /// - This is required by the IO Finalizer, and is cleared by it once used.
+    /// - Signers MUST reject PCZTs that contain `dummy_sk` values.
+    pub(crate) dummy_sk: Option<[u8; 32]>,
+
+    /// Proprietary fields related to the note being spent.
+    pub(crate) proprietary: BTreeMap<String, Vec<u8>>,
+}
+
+/// Information about an Orchard output within a transaction.
+#[serde_as]
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub(crate) struct Output {
+    //
+    // Output-specific Action effecting data.
+    //
+    // These are required fields that are part of the final transaction, and are filled in
+    // by the Constructor when adding an output.
+    //
+    pub(crate) cmx: [u8; 32],
+    pub(crate) ephemeral_key: [u8; 32],
+    /// The encrypted note plaintext for the output.
+    ///
+    /// Encoded as a `Vec<u8>` because its length depends on the transaction version.
+    ///
+    /// Once we have memo bundles, we will be able to set memos independently of Outputs.
+    /// For now, the Constructor sets both at the same time.
+    pub(crate) enc_ciphertext: Vec<u8>,
+    /// The encrypted note plaintext for the output.
+    ///
+    /// Encoded as a `Vec<u8>` because its length depends on the transaction version.
+    pub(crate) out_ciphertext: Vec<u8>,
+
+    /// The address that will receive the output.
+    ///
+    /// - This is set by the Constructor.
+    /// - This is required by the Prover.
+    #[serde_as(as = "Option<[_; 43]>")]
+    pub(crate) recipient: Option<[u8; 43]>,
```

Comment:
There's an annoyance here with the `zcash_client_backend` integration: we need to be able to get the recipient's UA, not just the Orchard component (and same for Sapling). We construct metadata values that carry this data along in the transaction construction process and normally we use it immediately, but with PCZT it needs to be carried along. I guess that this should go into the `proprietary` field; maybe that should be exposed via a mut getter so that it can be manipulated in that way?

Reply:
We'll also have to add proposal step metadata here. We can't add this metadata directly in the process of adding outputs to the `zcash_primitives` builder; instead, we have to track it separately, then mutate the resulting PCZT to add the metadata so that we can carry it over to where we store the transaction data.

Reply:
In earlier PCZT discussions this year I suggested that the transaction proposal should be extractable from the (set of) PCZT resulting from it, but got strong enough pushback against it that I dropped the idea. So yes, if you now do actually need to represent parts of the proposal in the PCZTs, you will need to do so via the `proprietary` field.

I think we will need to expose some kinds of getters and setters here, in particular for the Updater role (which I haven't implemented because it's so broad) because we need to be able to set `Zip32Derivation` values to inform the Keystone device of which keys to use.

Reply:
It's not actually part of the proposal per se; some of the metadata that we need to carry along is stuff that's derived from wallet state in the process of transaction construction. So I think that the decision to not make a `Proposal` extractable was actually correct, because it would have been incomplete with respect to what we need anyway.